### PR TITLE
fileless instances (issue #78) and instances with extra arguments

### LINKED ISSUE #78: Fileless Instances
We should add the possibility of fileless instances as an instance might be defined through input parameters ,e.g, algorithms that generate their data themselves and only need an input parameter like ``--seed=X``.

This might be done by adding an ``args`` key to instances and a new at-variable ``@INSTANCE_ARGS@`` that resolves to the respective instance args.

### PR DESCRIPTION
**Pull request by duylethanh**
This PR contains the following:
- restructuring/fixing the ``schemes/experiments.json`` schema and addition of ``extra_args``to the ``instances`` stanza
- support of fileless instances( resolves #78), by introducing the ``extra_args`` key for the ``instances``stanza
- documentation for fieless instances and the ``extra_args`` key

Below are some more details for the first two points:

The ``extra_args`` of instances are accessible via the already existing ``@EXTRA_ARGS@`` variable, which only resolved to the ``extra_args`` of variants before. Now it resolves to the concatenation of both (possible more, because we can have more than one variant for an experiment)``extra_args`` (from the instance and the variants).

-----

The validation schema for instances in the ``experiments.yml`` file can now roughly be expressed as:

```
if 'generator' in instance:
	check syntax for generator instances
if 'extensions' in instance:
	check syntax of multiple extension instances
if 'files' in instance:
	check syntax of arbitrary file instances
if neither of the above cases:
	check syntax for local/remote
```

(Notice the ``if-if``-construct instead of an ``if-elif``-construct which is due to the fact that [jsonschema 7.0](https://json-schema.org/understanding-json-schema/reference/conditionals.html) does not currently support ``if-elif``constructs.)

----

If an entity fails a ``oneOf`` construct, simexpal will now print out the errors for each subschema, e.g,: 

the entity

```
instances:
  - repo: local
    items:
      - name: baz
        extra_args: ['some','extra_args']
      # 'name' key is missing and 'foo' is an invalid key
      - extra_args: ['some','extra_args']
        foo: bar
  - repo: locjal    # typo of local
    items:
      - instance1
```

produces the following output when calling a ``simex`` command:

Screen Shot 2020-11-27 at 21 57 50

### CODE REVIEW

File: docs/experiments.rst [hunk 1]
```diff
@@ -103,6 +115,14 @@ as follows:
 The ``@INSTANCE:0@`` variable will resolve to ``/instance_directory/files[0]``, where ``files[0]`` is
 the first filename of the ``files`` key. Analogously for the ``@INSTANCE:1`` variable.
 
+If the instances have :ref:`InstanceExtraArguments`, we further need to add the :ref:`@-variable <AtVariables>`
+``@EXTRA_ARGS@`` to the experiment arguments, e.g,
+``args: ['./algorithm.py', '@INSTANCE:0', '@INSTANCE:1@', '@EXTRA_ARGS@']`` for the example above.
+``@EXTRA_ARGS@`` will resolve to the specified extra arguments of the respective instance (and also the used
+:ref:`variants <ExperimentsWithVariants>`, see below).
```
Comment (avdgrinten) [inline docs/experiments.rst:122]: I don't link that this paragraph is repeated three times. Can we shorten the repetitions and refer back to the first occurrence?

File: simexpal/util.py [hunk 2]
```diff
@@ -97,6 +97,19 @@ def _validate_dict(dictionary, source, schema_path):
 
 				print("simexpal: Validation error in {} at {}:\n{}\n{}".format(
 					source, err_source, err.instance, err.message), file=sys.stderr, end="\n\n")
+
+				# The error comes from subschemas in anyOf, oneOf or allOf.
+				if err.context:
```
Comment (avdgrinten) [inline simexpal/util.py:102]: `context` is only set for anyOf, oneOf and allOf? Did you test that is does not give unreasonable output in other cases?

File: simexpal/launch/common.py [hunk 3]
```diff
@@ -256,9 +262,11 @@ def compile_manifest(run):
 		'variants': variants_yml,
 		'revision': exp.revision.name if exp.revision else None,
 		'instance': run.instance.shortname,
-		'instance_filename': run.instance.yml_name,
+		'instance_yml_name': run.instance.yml_name,
```
Comment (avdgrinten) [inline simexpal/launch/common.py:265]: Why not just `instance_name`?

**Review by avdgrinten — COMMENTED** [review]
Looks mostly good, a few minor comments below.

File: simexpal/util.py (hunk 2 above)
Comment (duylethanh) [reply simexpal/util.py:102]: > `context` is only set for anyOf, oneOf and allOf? 

Yes, that should be the case. The following text comes from the [official documentation](https://python-jsonschema.readthedocs.io/en/stable/errors/): _With validators like anyOf, the ``ValidationError.context`` attribute can be used to see the sub-errors which caused the failure._

> Did you test that is does not give unreasonable output in other cases?

Yes, I did some testing and in other cases the ``context`` attribute is just an empty list ``[]``.

File: simexpal/launch/common.py (hunk 3 above)
Comment (duylethanh) [reply simexpal/launch/common.py]: I changed it to ``instance_name``

File: docs/experiments.rst (hunk 1 above)
Comment (duylethanh) [reply docs/experiments.rst]: I removed the repetition and references the first occurence

**Comment by duylethanh** [comment]
PR updated.

**Comment by avdgrinten** [comment]
LGTM.

**Pull request merged**